### PR TITLE
Redirect with regex

### PR DESCRIPTION
**Pull request by joverlee521**
## Description of proposed changes

Use `path-to-regexp` for dataset redirects. This allows us to use regex patterns to match datasets and redirect them
to a new pattern without having to list out all of the existing paths.

Motivated by https://github.com/nextstrain/nextstrain.org/pull/811#issuecomment-2033012211

## Checklist

- [x] Checks pass

### CODE REVIEW

File: src/redirects.js [hunk 1]
```diff
@@ -162,6 +162,12 @@ export function updateDatasetUrl(originalUrl) {
     ['/monkeypox/hmpxv1', '/mpox/clade-IIb'],
     ['/monkeypox/hmpxv1/big', '/mpox/lineage-B.1'],
     ["/monkeypox", "/mpox"],
+    /**
+     * We shifted avian-flu and seasonal-flu to top level to stop nesting them
+     * both under flu/ so that avian-flu is more discoverable
+     */
+    ['/flu/avian(.*)', '/avian-flu(.*)'],
+    ['/flu/seasonal(.*)', '/seasonal-flu(.*)'],
```
Comment (joverlee521) [inline src/redirects.js:170]: I'm including this mostly to demonstrate how we would redirect the flu paths.
I can drop it if we need to coordinate change with https://github.com/nextstrain/nextstrain.org/pull/811

File: src/redirects.js [hunk 2]
```diff
@@ -128,6 +127,58 @@ export const setup = (app) => {
 
 };
 
+/**
+ * Checks if the provided originalUrl matches are any of our datasetRedirectPatterns
+ * If the there is match, then return the new URL for the dataset.
+ * If there is not match, then return the original URL.
+ *
+ * @param {string} originalUrl
```
Comment (jameshadfield) [inline src/redirects.js:135]: This docstring should be changed to indicate that the input must be the pathname only (as the calling code does at the moment) - the function will return or destroy queries depending on if there's a `:version` match.

File: src/redirects.js (hunk 1 above)
Comment (jameshadfield) [reply src/redirects.js:170]: Yeah - nice to see how well these work! I'd say drop this commit from this PR and we can add it to a subsequent PR depending on the migration path we choose to take.

**Review by jameshadfield — CHANGES_REQUESTED** [review]
Looks like a good solution - thanks @joverlee521. Left a few comments inline. I'd favour merging this soonish as it'll be necessary for #811.

File: src/redirects.js (hunk 1 above)
Comment (joverlee521) [reply src/redirects.js]: Dropped https://github.com/nextstrain/nextstrain.org/commit/3e8319440a77b1ab7206f34c3fa7085546e33c21

File: src/redirects.js [hunk 3]
```diff
@@ -128,6 +127,58 @@ export const setup = (app) => {
...
+     * e.g. /monkeypox/hmpxv1/2022-09-04
+     */
+    ['/monkeypox/mpxv', '/mpox/all-clades'],
+    ['/monkeypox/hmpxv1', '/mpox/clade-IIb'],
+    ['/monkeypox/hmpxv1/big', '/mpox/lineage-B.1'],
+    ["/monkeypox", "/mpox"],
+  ];
+
+  // Include an optional versioned dataset pattern for all redirect patterns
+  const versionDatasetPattern = ":version(@{1}.*)?"
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply src/redirects.js]: Thanks! Updated in https://github.com/nextstrain/nextstrain.org/pull/812/commits/989b393ad1bca02ef53db6e420076fa801824911 to `:version(@.*)?`

File: src/redirects.js [hunk 4]
```diff
@@ -128,6 +127,58 @@ export const setup = (app) => {
...
+
+  for (const [originalPattern, redirectPattern] of datasetRedirectPatterns) {
+    const urlMatch = match(`${originalPattern}${versionDatasetPattern}`);
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply src/redirects.js]: Thanks! I need to get better at recognizing these patterns. 

Moved out of `updateDatasetUrl` in https://github.com/nextstrain/nextstrain.org/pull/812/commits/443e132a2e8958b03db8409e23aa72f8a8cd6d9c

File: src/redirects.js (hunk 2 above)
Comment (joverlee521) [reply src/redirects.js]: Updated docstring in https://github.com/nextstrain/nextstrain.org/pull/812/commits/989b393ad1bca02ef53db6e420076fa801824911

(I also updated the param to `originalUrlPathname` to be clearer)

Comment (jameshadfield) [reply src/redirects.js]: 👍

File: src/redirects.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jameshadfield) [reply src/redirects.js]: 👍

File: src/redirects.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameshadfield) [reply src/redirects.js]: Thanks!

**Review by jameshadfield — APPROVED** [review]

**Pull request merged**